source/agora/common/ManagedDatabase.d

```diff
@@ -120,6 +120,7 @@ public class ManagedDatabase
             {
                 // in most cases this would only trigger if there was a
                 // code-flow error
+                debug { import std.stdio : writefln; try { writefln!"================beginBatch failed:%s"(exc.message); } catch (Exception) {} }
```

Comment:
Why isn't the `log.fatal` message enough ?

Reply:
Did you actually encounter this error BTW ?

Reply:
On rollback yes. And the log.fatal was not printed as it called abort in the calling code. This was fixed but took some detective work.

Reply:
Hum, `abort` should give you a stacktrace. And I think a better fix for this would be to setup the assert handler in the unittests to print the logs.

Reply:
I will create a branch with this situation reproduced so we can take a better look at the scenario.

Reply:
It'd be good to debug the issue in itself. But what I meant is that the unittest runner should use [this](https://dlang.org/library/core/exception/assert_handler.html) so that the relevant log messages are printed when a thread `assert`s. This way you solve the issue everywhere.

Reply:
I created commit bfa93cc17c174646e7be465ef0463bc1cd792e1b with an example of where the rollback is called with no active transactions and the log.fatal is not printed. I will also take a look at adding an AssertionHandler to the tests.

Reply:
Changes are in https://github.com/bpfkorea/agora/pull/1273